Lazy

Denzed opened the pull request.

Denzed committed:
Implemented single threaded version of Lazy

Denzed committed:
Merge branch 'master' into Lazy

Denzed committed:
Implemented one test for single threaded Lazy implementation

Denzed committed:
Implemented remaining tests for single threaded Lazy implementation

Denzed committed:
Fixed null Supplier behaviour

Denzed committed:
Suddenly Travis CI failed because of CompileJava version. Trying fix

Denzed committed:
Once again

Denzed committed:
Just understood that Travis does not support openJDK 8

Denzed committed:
Implemented thread-safe Lazy

Denzed committed:
Updated build.gradle so it would log tests it runs

Denzed committed:
Almost finished

Denzed committed:
Fixing multi threaded versions

Denzed committed:
Refined JavaDoc and added (what in my opinion it is) standalone jar

Denzed commented:
По поводу того, что последний коммит не проверен Travis CI: https://www.traviscistatus.com/incidents/hmwq9yy5dh9d

jzuken commented:
мм, jar/Java 4th semester.jar зачем тут?

jzuken commented on src/main/java/lazy/LazyLockFree.java:
```diff
@@ -0,0 +1,34 @@
+ * @param <T> Return type
+ */
+class LazyLockFree<T> implements Lazy<T> {
+    private AtomicReference<Supplier<T>> supplierReference;
```
давайте лучше использовать AtomicReferenceFieldUpdater, оно меньше памяти использует (один на класс, а не на объект)

Denzed committed:
Fixes
Changes related to commented flaws
- Removed jar from the git repository
- Changed LazyLockFree in such a way that it now uses AtomicReferenceFieldUpdater instead of AtomicReference
- Rethought the behaviour in case of a null computation --- it now raises NullPointerException

jzuken commented:
бинарник всё ещё тут, -0.5

Denzed committed:
Finally removed jar

jzuken commented:
ок, 9.5

Denzed committed:
Whoops
